File: web/src/api/notification.ts
import type { Component } from 'vue'
import type { ApiListResponse } from './types'
import { AlertCircle, AlertTriangle, Bell, Clock, Info, UserCheck, UserPlus, UserX } from 'lucide-vue-next'
import { api } from './http'

// 通知相关类型定义
export interface Notification {
  id: string
  recipient_id: string
  type: string
  title: string
  content: string
  status: 'unread' | 'read'
  priority: 'low' | 'normal' | 'high' | 'urgent'
  related_id?: string
  related_type?: string
  read_at?: number
  expires_at?: number
  metadata?: string
  created_at: number
  updated_at: number
  is_expired?: boolean
}

export interface NotificationListResponse extends ApiListResponse<Notification> {
  unread_count: number
}

export interface NotificationStats {
  total_count: number
  unread_count: number
  read_count: number
  by_type: Record<string, number>
  by_priority: Record<string, number>
}

export interface CreateNotificationRequest {
  recipient_ids: string[]
  type: string
  title: string
  content: string
  priority?: 'low' | 'normal' | 'high' | 'urgent'
  expires_at?: number
}

export interface BulkNotificationRequest {
  user_roles?: string[]
  user_ids?: string[]
  type: string
  title: string
  content: string
  priority?: 'low' | 'normal' | 'high' | 'urgent'
  expires_at?: number
}

export interface NotificationTemplate {
  type: string
  title: string
  content: string
  priority: string
  description: string
  variables: string[]
}

export interface GetNotificationsRequest {
  page?: number
  page_size?: number
  status?: 'unread' | 'read' | 'all'
  type?: string
  priority?: string
}

// 通知API接口
export const notificationApi = {
  // 获取通知列表
  getNotifications: (params?: GetNotificationsRequest) => {
    return api.get<any, NotificationListResponse>('/notifications', { params })
  },

  // 标记通知为已读
  markAsRead: (id: string) => {
    return api.put(`/notifications/${id}/read`)
  },

  // 标记所有通知为已读
  markAllAsRead: () => {
    return api.put('/notifications/read-all')
  },

  // 获取未读通知数量
  getUnreadCount: () => {
    return api.get<any, { unread_count: number }>('/notifications/unread-count')
  },

  // 删除通知
  deleteNotification: (id: string) => {
    return api.delete(`/notifications/${id}`)
  },

  // 获取通知统计
  getStats: () => {
    return api.get<any, NotificationStats>('/notifications/stats')
  },

  // 创建通知（管理员）
  createNotification: (data: CreateNotificationRequest) => {
    return api.post('/notifications', data)
  },

  // 批量发送通知（管理员）
  sendBulkNotification: (data: BulkNotificationRequest) => {
    return api.post('/notifications/bulk', data)
  },

  // 获取通知模板
  getTemplates: () => {
    return api.get<NotificationTemplate[]>('/notifications/templates')
  },
}

// 通知类型映射
export const notificationTypeMap: Record<string, { icon: Component, label: string }> = {
  access_request_created: {
    icon: UserPlus,
    label: '新的访问申请',
  },
  access_request_approved: {
    icon: UserCheck,
    label: '访问申请已批准',
  },
  access_request_rejected: {
    icon: UserX,
    label: '访问申请已拒绝',
  },
  access_request_expired: {
    icon: Clock,
    label: '访问权限已过期',
  },
  secret_item_expiring: {
    icon: Clock,
    label: '密钥项即将过期',
  },
  secret_item_expired: {
    icon: Clock,
    label: '密钥项已过期',
  },
  security_alert: {
    icon: AlertCircle,
    label: '安全警报',
  },
}

// 通知优先级映射
export const notificationPriorityMap: Record<string, { label: string, color: string }> = {
  low: { label: '低', color: 'bg-gray-500' },
  normal: { label: '普通', color: 'bg-blue-500' },
  high: { label: '高', color: 'bg-orange-500' },
  urgent: { label: '紧急', color: 'bg-red-500' },
}

// 通知优先级图标
export const notificationPriorityIcons: Record<string, Component> = {
  low: Info,
  normal: Bell,
  high: AlertTriangle,
  urgent: AlertCircle,
}

File: api/types/notification.go
package types

import "github.com/akinoccc/hysaif/api/models"

// 通知相关请求类型

// GetNotificationsRequest 获取通知列表请求
type GetNotificationsRequest struct {
	Page     int    `form:"page" binding:"omitempty,min=1"`
	PageSize int    `form:"page_size" binding:"omitempty,min=1,max=100"`
	Status   string `form:"status" binding:"omitempty,oneof=unread read all"`          // unread, read, all
	Type     string `form:"type"`                                                      // 通知类型过滤
	Priority string `form:"priority" binding:"omitempty,oneof=low normal high urgent"` // 优先级过滤
}

// CreateNotificationRequest 创建通知请求（管理员使用）
type CreateNotificationRequest struct {
	RecipientIDs []string `json:"recipient_ids" binding:"required,min=1,dive,required"`
	Type         string   `json:"type" binding:"required,min=1,max=50"`
	Title        string   `json:"title" binding:"required,min=1,max=200"`
	Content      string   `json:"content" binding:"required,min=1,max=1000"`
	Priority     string   `json:"priority" binding:"omitempty,oneof=low normal high urgent"`
	ExpiresAt    uint64   `json:"expires_at"` // 过期时间，0表示不过期
}

// MarkNotificationRequest 标记通知请求
type MarkNotificationRequest struct {
	NotificationIDs []string `json:"notification_ids" binding:"required,min=1,dive,required"`
	Status          string   `json:"status" binding:"required,oneof=read unread"`
}

// NotificationListResponse 通知列表响应
type NotificationListResponse struct {
	Data        []models.Notification `json:"data"`
	Pagination  Pagination            `json:"pagination"`
	UnreadCount int64                 `json:"unread_count"`
}

// NotificationStatsResponse 通知统计响应
type NotificationStatsResponse struct {
	TotalCount  int64            `json:"total_count"`
	UnreadCount int64            `json:"unread_count"`
	ReadCount   int64            `json:"read_count"`
	ByType      map[string]int64 `json:"by_type"`
	ByPriority  map[string]int64 `json:"by_priority"`
}

// NotificationPreferencesRequest 通知偏好设置请求
type NotificationPreferencesRequest struct {
	EmailNotifications bool     `json:"email_notifications"`
	EnabledTypes       []string `json:"enabled_types"`
	QuietHours         struct {
		Enabled   bool   `json:"enabled"`
		StartTime string `json:"start_time" binding:"omitempty,datetime=15:04"` // HH:MM 格式
		EndTime   string `json:"end_time" binding:"omitempty,datetime=15:04"`   // HH:MM 格式
	} `json:"quiet_hours"`
}

// NotificationPreferencesResponse 通知偏好设置响应
type NotificationPreferencesResponse struct {
	UserID             string   `json:"user_id"`
	EmailNotifications bool     `json:"email_notifications"`
	EnabledTypes       []string `json:"enabled_types"`
	QuietHours         struct {
		Enabled   bool   `json:"enabled"`
		StartTime string `json:"start_time"`
		EndTime   string `json:"end_time"`
	} `json:"quiet_hours"`
	UpdatedAt uint64 `json:"updated_at"`
}

// BulkNotificationRequest 批量通知请求
type BulkNotificationRequest struct {
	UserRoles []string `json:"user_roles" binding:"omitempty,dive,oneof=super_admin sec_mgr dev auditor"` // 按角色发送
	UserIDs   []string `json:"user_ids" binding:"omitempty,dive,required"`                                // 指定用户ID
	Type      string   `json:"type" binding:"required,min=1,max=50"`
	Title     string   `json:"title" binding:"required,min=1,max=200"`
	Content   string   `json:"content" binding:"required,min=1,max=1000"`
	Priority  string   `json:"priority" binding:"omitempty,oneof=low normal high urgent"`
	ExpiresAt uint64   `json:"expires_at"`
}

// NotificationTemplateResponse 通知模板响应
type NotificationTemplateResponse struct {
	Type        string   `json:"type"`
	Title       string   `json:"title"`
	Content     string   `json:"content"`
	Priority    string   `json:"priority"`
	Description string   `json:"description"`
	Variables   []string `json:"variables"` // 模板变量列表
}
